pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [duplicate constraint name]

pkg/ccl/testccl/workload/schemachange/schemachange_test.TestWorkload [failed](https://mesolite.cluster.engflow.com/invocations/default/6e57e876-68f1-47c8-961a-924108302686?testReportRun=1&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL3Rlc3RjY2wvd29ya2xvYWQvc2NoZW1hY2hhbmdlOnNjaGVtYWNoYW5nZV90ZXN0) on release-24.1 @ [82173b48203cab71165d9552b7ab39606da37b72](https://github.com/cockroachdb/cockroach/commits/82173b48203cab71165d9552b7ab39606da37b72):


```
    "result": false
   },
   {
    "query": "SELECT count(*) FROM schema_w1_24.table_w0_55",
    "result": 0
   }
  ],
  "previousStatements": [
   "ALTER TABLE schema_w1_24.table_w0_55 ADD CONSTRAINT \"table_w1_126_co%pl126_w1_138_table_w0_55_col55_w0_68_fk\" FOREIGN KEY (col55_w0_68) REFERENCES schema_w1_89.table_w1_126 (\"co%pl126_w1_138\") ON DELETE CASCADE ON UPDATE CASCADE"
  ],
  "usesDeclarativeSchemaChanger": true
 }
}
    schema_change_external_test.go:129: 
        	Error Trace:	pkg/ccl/testccl/workload/schemachange/schemachange_test_test/pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:129
        	Error:      	Received unexpected error:
        	            	***UNEXPECTED ERROR; Received an unexpected execution error.: ERROR: duplicate constraint name: "table_w1_126_co%pl126_w1_138_table_w0_55_col55_w0_68_fk" (SQLSTATE 42710)
        	            	(1) forced error mark
        	            	  | "fatal error when running txn"
        	            	  | github.com/cockroachdb/errors/withstack/*withstack.withStack::
        	            	Wraps: (2)
        	            	Wraps: (3) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*opStmt).executeStmt
        	            	  | 	github.com/cockroachdb/cockroach/pkg/workload/schemachange/operation_generator.go:3191
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).runInTxn
        	            	  | 	github.com/cockroachdb/cockroach/pkg/workload/schemachange/schemachange.go:514
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).run
        	            	  | 	github.com/cockroachdb/cockroach/pkg/workload/schemachange/schemachange.go:608
        	            	  | pkg/ccl/testccl/workload/schemachange/schemachange_test_test.TestWorkload.TestWorkload.func4.func6
        	            	  | 	pkg/ccl/testccl/workload/schemachange/schemachange_test_test/pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:118
        	            	  | golang.org/x/sync/errgroup.(*Group).Go.func1
        	            	  | 	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:78
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1695
        	            	Wraps: (4) ***UNEXPECTED ERROR; Received an unexpected execution error.
        	            	Wraps: (5) ERROR: duplicate constraint name: "table_w1_126_co%pl126_w1_138_table_w0_55_col55_w0_68_fk" (SQLSTATE 42710)
        	            	Error types: (1) *markers.withMark (2) *schemachange.ErrorState (3) *withstack.withStack (4) *errutil.withPrefix (5) *pgconn.PgError
        	Test:       	TestWorkload
Schema Workload Stats
Total Schema Statements Executed = 95
Total Schema Statements Succeeded = 95
Total Schema Statement Expected Failures = 0
Total Transactions Committed = 45
Total Transactions Rolled Back = 78
Total Transactions Executed = 123
    schema_change_external_test.go:95: backup, tracing data, and system table dumps in /var/lib/engflow/worker/work/2/exec/bazel-out/k8-fastbuild/testlogs/pkg/ccl/testccl/workload/schemachange/schemachange_test/test.outputs/logTestWorkload1827159682
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestWorkload1827159682
--- FAIL: TestWorkload (12.14s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=1</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #138261 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [check for columnHasSingleUniqueConstraint needs to handle hash-sharded primary key]  [C-test-failure O-robot P-2 T-sql-foundations branch-release-24.3]
- #137979 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [descriptor not found during DROP VIEW commit] [C-test-failure O-robot P-3 T-sql-foundations X-duplicate branch-release-23.2.19-rc]
- #137972 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [requested column not in index] [C-test-failure O-robot P-2 T-sql-foundations branch-release-24.3.3-rc]
- #136966 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [duplicate key value violates unique constraint] [C-test-failure O-robot P-2 T-sql-foundations branch-master]
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWorkload.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-46163

pkg/ccl/testccl/workload/schemachange/schemachange_test.TestWorkload [failed](https://tanzanite.cluster.engflow.com/invocations/default/106249aa-b305-49c6-9b87-356ebd91b0d9?testReportRun=2&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL3Rlc3RjY2wvd29ya2xvYWQvc2NoZW1hY2hhbmdlOnNjaGVtYWNoYW5nZV90ZXN0) on release-24.1 @ [d2577d5594576db7fb051ac40a3801e61222a259](https://github.com/cockroachdb/cockroach/commits/d2577d5594576db7fb051ac40a3801e61222a259):


```
    "result": false
   },
   {
    "query": "SELECT count(*) FROM schema_w1_96.table_w0_104",
    "result": 0
   }
  ],
  "previousStatements": [
   "ALTER TABLE schema_w1_96.table_w0_104 ADD CONSTRAINT table_w0_123_col123_w0_125_table_w0_104_col104_w0_113_fk FOREIGN KEY (col104_w0_113) REFERENCES schema_w0_73.table_w0_123 (col123_w0_125) ON DELETE CASCADE ON UPDATE CASCADE"
  ],
  "usesDeclarativeSchemaChanger": true
 }
}
    schema_change_external_test.go:129: 
        	Error Trace:	pkg/ccl/testccl/workload/schemachange/schemachange_test_test/pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:129
        	Error:      	Received unexpected error:
        	            	***UNEXPECTED ERROR; Received an unexpected execution error.: ERROR: duplicate constraint name: "table_w0_123_col123_w0_125_table_w0_104_col104_w0_113_fk" (SQLSTATE 42710)
        	            	(1) forced error mark
        	            	  | "fatal error when running txn"
        	            	  | github.com/cockroachdb/errors/withstack/*withstack.withStack::
        	            	Wraps: (2)
        	            	Wraps: (3) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*opStmt).executeStmt
        	            	  | 	github.com/cockroachdb/cockroach/pkg/workload/schemachange/operation_generator.go:3191
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).runInTxn
        	            	  | 	github.com/cockroachdb/cockroach/pkg/workload/schemachange/schemachange.go:514
        	            	  | github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWorker).run
        	            	  | 	github.com/cockroachdb/cockroach/pkg/workload/schemachange/schemachange.go:608
        	            	  | pkg/ccl/testccl/workload/schemachange/schemachange_test_test.TestWorkload.TestWorkload.func4.func6
        	            	  | 	pkg/ccl/testccl/workload/schemachange/schemachange_test_test/pkg/ccl/testccl/workload/schemachange/schema_change_external_test.go:118
        	            	  | golang.org/x/sync/errgroup.(*Group).Go.func1
        	            	  | 	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:78
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1695
        	            	Wraps: (4) ***UNEXPECTED ERROR; Received an unexpected execution error.
        	            	Wraps: (5) ERROR: duplicate constraint name: "table_w0_123_col123_w0_125_table_w0_104_col104_w0_113_fk" (SQLSTATE 42710)
        	            	Error types: (1) *markers.withMark (2) *schemachange.ErrorState (3) *withstack.withStack (4) *errutil.withPrefix (5) *pgconn.PgError
        	Test:       	TestWorkload
Schema Workload Stats
Total Schema Statements Executed = 97
Total Schema Statements Succeeded = 97
Total Schema Statement Expected Failures = 0
Total Transactions Committed = 52
Total Transactions Rolled Back = 81
Total Transactions Executed = 133
    schema_change_external_test.go:95: backup, tracing data, and system table dumps in /var/lib/engflow/worker/work/3/exec/bazel-out/k8-fastbuild/testlogs/pkg/ccl/testccl/workload/schemachange/schemachange_test/run_2_of_25/test.outputs/logTestWorkload1288743755
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestWorkload1288743755
--- FAIL: TestWorkload (11.29s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=2</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #138721 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [C-test-failure O-robot P-3 T-sql-foundations branch-release-23.1]
- #137972 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [requested column not in index] [C-test-failure O-robot P-2 T-sql-foundations branch-release-24.3.3-rc]
- #136966 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [duplicate key value violates unique constraint] [C-test-failure O-robot P-2 T-sql-foundations branch-master]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWorkload.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/workload/schemachange/error_screening.go
+++ b/pkg/workload/schemachange/error_screening.go
@@ -1011,11 +1011,13 @@ SELECT COALESCE(
 }
 
 func (og *operationGenerator) constraintExists(
-	ctx context.Context, tx pgx.Tx, constraintName string,
+	ctx context.Context, tx pgx.Tx, tableName, constraintName tree.Name,
 ) (bool, error) {
+	// Note: information_schema.table_constraints contains constraints that are
+	// in the dropping state, but pg_constraint.constraints does not.
 	return og.scanBool(ctx, tx, `SELECT EXISTS(
-		SELECT * FROM pg_catalog.pg_constraint WHERE conname = $1
-	 )`, constraintName)
+		SELECT * FROM information_schema.table_constraints WHERE table_name = $1 AND constraint_name = $2
+	 )`, string(tableName), string(constraintName))
 }
 
 func (og *operationGenerator) rowsSatisfyFkConstraint(

--- a/pkg/workload/schemachange/operation_generator.go
+++ b/pkg/workload/schemachange/operation_generator.go
@@ -367,7 +367,7 @@ func (og *operationGenerator) addUniqueConstraint(ctx context.Context, tx pgx.Tx
 	if err != nil {
 		return nil, err
 	}
-	constraintExists, err := og.constraintExists(ctx, tx, constraintName)
+	constraintExists, err := og.constraintExists(ctx, tx, tableName.ObjectName, tree.Name(constraintName))
 	if err != nil {
 		return nil, err
 	}
@@ -890,7 +890,7 @@ func (og *operationGenerator) addForeignKeyConstraint(
 	if err != nil {
 		return nil, err
 	}
-	constraintExists, err := og.constraintExists(ctx, tx, string(constraintName))
+	constraintExists, err := og.constraintExists(ctx, tx, childTable.ObjectName, constraintName)
 	if err != nil {
 		return nil, err
 	}

